yarl/_url.py

```diff
@@ -925,6 +930,48 @@ def _make_netloc(
             ret = user + "@" + ret
         return ret
 
+    @classmethod
+    @lru_cache  # match the same size as urlsplit
+    def _split_netloc(
+        cls,
+        netloc: str,
+    ) -> Tuple[Optional[str], Optional[str], str, Optional[int]]:
+        """Split netloc into username, password, host and port.
+
+        host is always encoded
+        """
+        if "@" not in netloc:
+            username: Optional[str] = None
+            password: Optional[str] = None
+            hostinfo = netloc
+        else:
+            userinfo, _, hostinfo = netloc.rpartition("@")
+            username, have_password, password = userinfo.partition(":")
+            if not have_password:
+                password = None
+
+        if "[" in hostinfo:
+            _, _, bracketed = hostinfo.partition("[")
+            hostname, _, port_str = bracketed.partition("]")
+            _, _, port_str = port_str.partition(":")
+        else:
+            hostname, _, port_str = hostinfo.partition(":")
+
+        if not hostname:
+            raise ValueError("Invalid URL: host is required for absolute urls")
+
+        if not port_str:
+            port: Optional[int] = None
+        else:
+            try:
+                port = int(port_str)
+            except ValueError:
+                raise ValueError("Invalid URL: port can't be converted to integer")
+            if not (0 <= port <= 65535):
+                raise ValueError("Port out of range 0-65535")
+
+        return username, password, cls._encode_host(hostname), port
```

Comment:
Might make more sense to move the encode to __new__ since it's the only place we this. Than we can reuse this function as a basis to parse when one of the raw netloc properties is accessed so we can lazy fill the cache for the pre encoded case